### ISSUE TITLE
Implement Label in Wechaty

### ISSUE THREAD

**Issue by su-chang** [wechaty/wechaty#1856]
**Is your feature request related to a problem? Please describe.**

if someone has a lot of friends, `Label` is indeed needed for manage friends list.

**Describe the solution you'd like**
we need some function like these: 
  - get all labels (For example: [{label: 'xxx', members: ['xxx1', 'xxx2', ... , 'xxxx']}])
  - add label
  - update friend's label

**Describe alternatives you've considered**

The function of `Label` is depended on puppet which one has supported.

- supported
  - macpro
  - padplus
- not supported
  - puppeteer
  - padpro

if the puppet is not supported, we can do nothing.

**Additional context**
I will create a PR for this issue later.

[enhancement]

**Comment by huan** [comment]
Thank you very much for the proposal!

A PR will be welcome for discussing, and also please notice that we have another thread talked about this topic before.

Please link to that issue and also read it carefully to learn more from the past discussions.

**Comment by su-chang** [comment]
#1627

**Comment by su-chang** [comment]
#1627

**Comment by huan** [comment]
Glad to see you had connected to the previous conversion.

It is very important that we should always continue our discussion, based on the already learned knowledge, instead of creating a branding new one from scratch.

**Comment by su-chang** [comment]
@huan 
Here are some Raw API and their response data.

- Raw API name : AddLabel

> bot.Tag.newTag(tag: string)

```js
{
  labelList:
    [ { LabelName: '测试标签1', LabelID: 1 }, ],
  loginer: 'xxx',
  count: 1,
  message: '',
  status: 0,
}
```

- Raw API name : GetAllLabel
> bot.Tag.tagList()
```js
{
  labelList:
   [ { LabelName: '测试标签1', LabelID: 1 },
     { LabelName: '测试标签3', LabelID: 3 },
     { LabelName: '测试标签2', LabelID: 2 } ],
  loginer: 'xxx',
  count: 3,
  message: '',
  status: 0
}
```

- Raw API name : ModifyLabel
> contact.addTag(contactId:string, tag: string)
> we can get `contactId` by contact
```js
{
  loginer: 'xxx',
  labelIds: '1,3',
  message: '',
  userName: 'Soul001001',
  status: 0
}
```

**Comment by su-chang** [comment]
For wechaty:

I think we should create a `Tag` class, the methods list:

- newTag(tag: string): Promise<string>
- findAll(): Promise<TagPayload []>
- find(tag: string): Promise<TagPayload>
- sync(): Promise<void>

> maybe we also should have these methods, but the underlying WeChat Raw API does not support directly.

- updateTag(oldTag: string, newTag: string): Promise<void>
> we could implement this method with other methods, but not handy.

and add some methods on `Contact` class, the methods list:

- addTag(tag: string): Promise<string>
- tags(): Promise<TagPayload []>
- delTag(): Promise<void>
- delTag(tag: string): Promise<void>

@huan  I am willing to get some advices from your side. Thank you.

**Comment by huan** [comment]
Cloud you show some design example for:

1. How can we add a tag to a contact?
1. How can we search for contacts by tag?

**Comment by su-chang** [comment]
const contact = bot.Contact.load('xxx')
contact.addTag('Test') // for 1
bot.Tag.getContactList('Test') // for 2, this method is not implemented.

Maybe we should change TagPayload -> Tag ?
```js
TagPayload {
  tagId: string,
  tagName: string,
}
```

**Comment by huan** [comment]
1. About add a tag, my design would like to be:

```ts
const tag = bot.Tag.get('myTag')
tag.set(contact)
tag.del(contact)
tag.set(favourite) // a Favorite instance, in the future
```

2. about the search, I'd like to be:

```ts
const myTag = bot.Tag.get('myTag')
const contactList = bot.Contact.findAll({ tag: myTag })
```

I'm open for discussion, the goal is to keep our API simple, clear, and concise.

**Comment by su-chang** [comment]
> 1. About add a tag, my design would like to be:
> 
> ```ts
> const tag = bot.Tag.get('myTag')
> tag.set(contact)
> tag.del(contact)
> ```

for this design, I think its better:
```js
const tag = bot.Tag.get('myTag')
tag.set(contactList)
tag.del(contactList)
tag.del() // this method for delete all contacts from tag
```

so we manipulate contact's tag by Tag, rather than Contact?

**Comment by huan** [comment]
I feel to manipulate the contact tag by Tag will be better, because the tag might be reused by other types of items, like a Favorite.

BTW: I do not think we should provide a `tag.del()` to delete all the items that have that tag, it's too dangerous. A developer might call this method by mistake and then he lost all his data.

**Comment by su-chang** [comment]
OK, agree!

BTW, I have no idea about the Favorite.

it can add some Image / Video / UrlLink or something else to one's Favorite?

right?

**Comment by huan** [comment]
Yes, what you think about the Favorite is all right. It should just like what we can do in Wechat App.

**Comment by su-chang** [comment]
#### Tag Side Methods
- create a new Tag
  ```typescript
  const tag = bot.Tag.create('newTag')
  ```

- delete a Tag  `Not Supported By API`
  ```typescript
  bot.Tag.del(tag)
  ```

- update a Tag  `Not Supported By API`
  ```typescript
  const tag = bot.Tag.update(oldTag, newTag)
  ```

- get a Tag, it will return TagPayload
  ```typescript
  const tag = bot.Tag.get('newTag')
  ```

- get all tags across the Bot
  ```typescript
  const tagList = bot.Tag.findAll()
  ```

#### Contact Side Methods

- get all tags of the Contact
  ```typescript
  const contactTagList = contact.tags()
  ```

#### Double Side Methods
- get all contacts of the Tag
  ```typescript
  const contactList = bot.Contact.findAll({tag: 'myTag'})
  ```
  > or maybe it can be implements like this way:

  ```typescript
  const tag = bot.Tag.get('myTag')
  const contactList = tag.contactList()
  ```

- set Contact to Tag
  ```typescript
  const tag = bot.Tag.get('myTag')
  tag.set(contact)
  ```

- set Tag to Contact
  ```typescript
  const tag = bot.Tag.get('myTag')
  contact.setTag(tag)
  ```
  or
  ```typescript
  contact.setTag('myTag')
  ```
  > if the Tag not existed, we should create a new Tag first

- remove Contact from Tag
  ```typescript
  tag.del('myTag')
  ```

- remove Tag from Contact
  ```typescript
  contact.delTag('myTag')
  ```


Maybe manipulate Tag and Contact by each other will be more convenient? Willing to get your suggestions. Thank you.

BTW: for Favorite, the label is different between Contact and Favorite, So are you sure the Tag could be reused by Favorite?

**Comment by huan** [comment]
My opinion would be the following methods should not exist in our design:

1. ~~Tag.create()~~
1. ~~Tag.del~~
1. ~~Tag.update()~~
1. ~~Contact.setTag()~~
1. ~~Contact.delTag()~~

Because they are more like to be confusing the developers instead of being useful.

And, Favorite and Contact is different, but the Tag abstract should be the same.

Please review my previous design suggestions, and try to re-think your design.

I'll review your PR of the wechaty-puppet later. (I feel it needs to be changed a lot after a quick look at it)

**Comment by su-chang** [comment]
Thanks for your suggestions. @huan 

The new design:

```js
TagPayload {
  id: string,
  name: string,
}
```
- Tag Class
```js
1. Tag.create(tag: string): Promise<TagPayload>
2. Tag.add(to: Contact): Promise<void>
3. Tag.del(from: Contact): Promise<void>
4. Tag.findAll(): Promise<TagPayload>
```

- Contact Class
```js
1. contact.tags(): Promise<TagPayload []>
```

That's all new methods related to the Label.

**Comment by huan** [comment]
`Tag.create()` should be `Tag.get()` because no one care about creating any tags, what we need is a tag with a name.

P.S. I have written a design for you before at https://github.com/Chatie/wechaty/issues/1627#issuecomment-542234015, please review it before you go any further.

**Comment by su-chang** [comment]
Of course, I have reviewed them. But I think `create` is better than `get` in semantic. Here are some reasons about this:

`create` means that we will get a new Label that we want, so user will know what they do.

`get` means that we will find a Label which named xxx. If we use `get` to create a new Label, any string will be created as a new tag, and then it will make a lot of new Labels. But the bot can not delete Label.

Do you agree with me ?

**Comment by huan** [comment]
I'm sorry I can not agree with you.

Wechat has neither any UI for creating tags (itself) nor any UI for deleting tags (itself), hasn't it?

**Comment by su-chang** [comment]
Nope, WeChat App has `Create Tag` and `Delete` for tag in Tags.

**Comment by huan** [comment]
I don't think so. Could you please provide some evidence to prove your opinions?

**Comment by su-chang** [comment]
You can prove it in your phone easily.

WeChat App -> Contacts -> Tags -> `EVIDENCE`

P.S. Something went really wrong, that I can not upload image here.

**Comment by huan** [comment]
That's not my point.

Everything will be ok without this feature. You can image that and then will agree with me. We can totally ignore it without any problem.

To make my point clear: there's no need to "create" a tag explictly. Just use it should be the best practice.

**Comment by su-chang** [comment]
Got it.

I'm just afraid of someone will use `get` to search Tag, and then will make a lot of Tags which are useless, and the terrible thing is that they can not delete them by bot(*NOT SUPPORTED BY API*), they need to delete them one by one in their WeChat App. 

If the user could use `get` like what we want, there is no problem with `get` to create a new Tag.

**Comment by windmemory** [comment]
@huan @su-chang Any updates on this issue?

**Comment by su-chang** [comment]
@windmemory  I will change some related methods in PR later, the best practice is better.

**Comment by huan** [comment]
Implemented by #1864
